src/ng/q.js

```diff
@@ -343,6 +343,14 @@ function qFactory(nextTick, exceptionHandler, errorOnUnhandledRejections) {
     }
   });
 
+  function PassToExceptionHandlerError(message) {
+    this.name = '$$PassToExceptionHandlerError';
+    this.message = message;
+    this.stack = (new Error()).stack;
+  }
```

Comment:
"Subclassing" Error is actually tricky. For example, different browsers use different properties (such as `line`, `sourceId`, `stackArray`). Other browsers do not populate the `stack` property until the error is actually thrown (so `(new Error).stack` will be empty). And finally, the stack trace will be a little confusing by showing a slightly different location:

```
# Expected
Error: Some error message
    at <where-the-error-was-actually-thrown>:<line>:<column>
    ...

# Actual
Error
    at PassToExceptionHandlerError (<file>:<line>:<column>)
    at <where-the-error-was-actually-thrown>:<line>:<column>
    ...
```

Here is what we do in `ngMock` to mitigate these issues:

https://github.com/angular/angular.js/blob/ecc09a45f02454e4e2478ad22cd0d179dccc7916/src/ngMock/angular-mocks.js#L3109-L3118

https://github.com/angular/angular.js/blob/ecc09a45f02454e4e2478ad22cd0d179dccc7916/src/ngMock/angular-mocks.js#L3122-L3128

https://github.com/angular/angular.js/blob/ecc09a45f02454e4e2478ad22cd0d179dccc7916/src/ngMock/angular-mocks.js#L3162

##
Another idea would be wrap the error in a non-error object and unwrap it in `$q`:

**src/ng/q.js**
```diff
+$Q.$$PassToExceptionHandlerWrapper = function PassToExceptionHandlerWrapper(error) { this.error = error; }
 ...
 function processQueue(state) {
   ...
   } catch (e) {
+    var passToExceptionHandler = e && e instanceof PassToExceptionHandlerWrapper;
+    e = passToExceptionHandler ? e.error : e;
     rejectPromise(promise, e);
+    if (passToExceptionHandler) {
+      exceptionHandler(e);
+    }
   }
```

**src/ngMock/angular-mocks.js**
```diff
+try {
+  // Throw the error to ensure the `stack` property is populated.
   throw wasExpected ?
       new Error('No response defined !') :
       new Error('Unexpected request: ' + method + ' ' + url + '\n' +
           (expectation ? 'Expected ' + expectation : 'No more request expected'));
+} catch (err) {
+  throw new $q.$$PassToExceptionHandlerWrapper(err);
+}
```

Then again, @mgol will say this violates the Promises/A+ spec and he will be right (since we will be rejecting with a different object than the one that was thrown) 😟 

##
Given all that, I think I might prefer the original "custom `$$passToExceptionHandler` property on the error" approach 😞 

What do others think?

Reply:
Are these issues with custom Errors still valid on the browser we support? I thought they were more IE9ish??

Reply:
The "no stack until thrown" issue is IE10+ (and PhantomJS which is not officially supported).
I am not sure about the extra properties, but iirc they are Safari-specific (could be wrong though).
The "extra line in stack trace" issue is by design, but should be easy to work around.

Basically all the issues are "work-around-able". The question is, is it worth the extra complexity?
<sub>(I don't feel strongly about it, though. Happy to go either way.)</sub>

Reply:
So there are 3 options:
1. Field on normal error (previous solution
2. Class inheriting from error (current code in PR, plus maybe changes to the way class defined)
3. Class wrapping error

Which way should be implemented?